### DISCUSSION TITLE
Recurring Donation with Custom Payment Gateway

### DISCUSSION THREAD

**Discussion by dightinfotechcom** [impress-org/givewp#7049] [Developer Q&A]
HI Dev Team!

I created a plugin as addon for including custom Payment gateways in WPGive Donation Plugin. My Plugin is working Fine for One Time Payment. But I don't get hook to use recurring payment with my custom Payment Gatway. I went through the documentation of GiveWP.  But Ididn't find what I want. 

````
<?php

// Get the Easy Merchant payment gateway.
$easy_merchant_gateway = givewp_recurring_donations_get_gateway( 'easy_merchant' );

// Create a new recurring donation subscription.
$subscription = $easy_merchant_gateway->create_subscription( array(
    'amount' => 10,
    'billing_interval' => 'monthly',
    'billing_start_date' => date( 'Y-m-d' ),
    'donor_name' => 'John Doe',
    'donor_email' => 'john.doe@example.com',
) );

// Update an existing recurring donation subscription.
$existing_subscription = $easy_merchant_gateway->get_subscription( 12345 );
$existing_subscription->update( array(
    'amount' => 20,
    'billing_interval' => 'yearly',
) );

// Cancel a recurring donation subscription.
$easy_merchant_gateway->cancel_subscription( 12345 );

// Process a payment for a recurring donation subscription.
$payment = $easy_merchant_gateway->process_payment( $subscription );

// Get the status of a recurring donation subscription.
$status = $easy_merchant_gateway->get_subscription_status( 12345 );

// Get the history of a recurring donation subscription.
$history = $easy_merchant_gateway->get_subscription_history( 12345 );

?>
````

After creating above file, I included in pluign main file. But I got errot of "givewp_recurring_donations_get_gateway" is not define. I took some help and in plugin main file, I write these two line of code

```
define('GIVEWP_RECURRING_DONATIONS_API_VERSION', '2.0');
define('GIVEWP_RECURRING_DONATIONS_API_PATH', plugins_url('includes/api', dirname(__FILE__)));
```

I am not sure which file is need to put in my plugin file, to use hooks and class, function from addons Give - Recurring Donations plugin.

**Comment by jonwaldstein** [comment]
Hey @dightinfotechcom,

I would suggest starting by creating a Gateway using our gateway api from [this](https://givewp.com/documentation/developers/how-to-build-a-gateway-add-on-for-givewp/) documentation.

We are still working on the docs for recurring donations but if you have GIve - Recurring Donations, you can look at how other gateways like `TestGatewaySubscriptionModule` are being created and loaded as a `SubscriptionModule` using the filter `"givewp_gateway_{gatewayId}_subscription_module"`.

**Reply by dightinfotechcom** [reply]
Thank you for replying back @jonwaldstein 
Somehow I manage for recurring donation with some static data in test mode. But I want to know how can I get card details like cardholdername, number, expiry month, year and cvv from custom paymentgateway form in PHP to post data on custom payment gateways api? 

`function give_get_donation_easy_cc_info()
{

  // Sanitize the values submitted with donation form.
  $post_data = give_clean($_POST); // WPCS: input var ok, sanitization ok, CSRF ok.

  $cc_info                   = [];
  $cc_info['cardholder_name']      = !empty($post_data['cardholder_name']) ? $post_data['cardholder_name'] : '';
  $cc_info['card_number']          = !empty($post_data['card_number']) ? $post_data['card_number'] : '';
  $cc_info['cvc']                  = !empty($post_data['cvc']) ? $post_data['cvc'] : '';
  $cc_info['exp_month']            = !empty($post_data['exp_month']) ? $post_data['exp_month'] : '';
  $cc_info['exp_year']             = !empty($post_data['exp_year']) ? $post_data['exp_year'] : '';

  // Return cc info.
  return $cc_info;
}`
I am trying like this and I want to use the card details in other function but couldn't get it. What thing I am missing? Please have a look.

**Reply by dightinfotechcom** [reply]
I got the credit card info. using the same function. but now issues is I am getting last 4 digit of card number only. While I need 16 digit number to POST on API. So I could get proper response. Please help.

**Reply by jonwaldstein** [reply]
Hey @dightinfotechcom a few questions:
1. Are you using the new gateway api I mentioned?
2. Are you able to share a repo with me to look further into your implementation?
3. Are you familiar with React js at all?

Payment Gateways can be complicated integrations to build and also explain how to build.  We definitely want to empower third party integrations like yourself but it can be difficult without actually seeing the full integration. Although we have some basic documentation on how to build a gateway for GiveWP, complexity is situational.  After answering those questions I can determine the best way to help you move forward which I anticipate will be collaborative.

**Reply by dightinfotechcom** [reply]
Hey @jonwaldstein 
Answer:
1.  No, I have all custom code using some hooks of GiveWP.
2. Yes, I can share the repo of my plugin to look into it. 
3. Yes I am familiar with React JS

I have invite you for collaboration. Please look the invitation. 
You can reach me anytime on mail: arvind.kumar@dightinfotech.com for any queries.

**Reply by jonwaldstein** [reply]
@dightinfotechcom thank you for the answers and inviting me to the repo, I will take a look.  However, the only way we recommend building custom gateways with GiveWP, is with our gateway api.  It is much more stable, easier to use and will require a lot less code.  I would highly suggest reading the docs and would be happy to help you with the migration https://givewp.com/documentation/developers/how-to-build-a-gateway-add-on-for-givewp/

**Reply by dightinfotechcom** [reply]
Sure, I will go through it and will update my code, if needed.Currently I have used few hooks only. I need one time and refurring payment both.

**Reply by jonwaldstein** [reply]
Sounds good! Just to clarify, using the gateway api is the only way to make a gateway compatible with GiveWP 3.0+ donation forms. Using hooks will only work for GiveWP v2 forms

**Reply by dightinfotechcom** [reply]
Thank you Jon. Do I need to start over my plugin from scratch to make my  gateway compatible with GiveWP 3.0+ donation form? Did you checked the repo? I invited you as a collaborator.

**Reply by jonwaldstein** [reply]
@dightinfotechcom yes that would be the only way.  I was able to scaffold a PR in your repo to help get started.  This could be used as a starting point.  Of course, the specific details of the gateway api and flow will need to implemented https://github.com/dightinfotechcom/easymerchant-givewp/pull/1

**Reply by dightinfotechcom** [reply]
@jonwaldstein Thank you! I merged the changes you made and I tried on development server. But I am getting this error
"Error: Sorry we're unable to process subscription donations with EasyMerchant. Please switch to the Stripe - Credit Card payment gateway to process a subscription donation."

I used easymerchan gateway api to get the charge_id

```
private function makePaymentRequest(array $data): array
    {
        $cc_info                  = give_get_donation_easymerchant_cc_info();
        // Use the card details in this function calling from "give_get_donation_easymerchant_cc_info" function.
        $cc_holder              = $cc_info['card_name'];
        $cc_number              = $cc_info['card_number_easy'];
        $cardNumber             = str_replace(' ', '', $cc_number);
        $month                  = $cc_info['card_exp_month'];
        $year                   = $cc_info['card_exp_year'];
        $cc_cvc                 = $cc_info['card_cvc'];
        $currentDate       = date("m/d/Y");
        $originalValues         = ["day", "week", "month", "quarter", "year"]; // API support these terms
        $replacementValues      = ["daily", "weekly", "monthly", "quarterly", "yearly"]; //givewp support these terms
        if (isset($data['period'])) {
            $originalValue        = $data['period'];
            $key                  = array_search($originalValue, $originalValues);
            if ($key !== false) {
                $data['period'] = $replacementValues[$key];
            }
        }
        $curl = curl_init();
        curl_setopt_array($curl, array(
            CURLOPT_URL => 'https://stage-api.stage-easymerchant.io/api/v1/charges/',
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => json_encode([
                'payment_mode'   => 'test',
                'amount'         => $data['amount'],
                'name'           => $data['user_info']['first_name'],
                'email'          => $data['user_email'],
                'description'    => 'test',
                'start_date'     => $currentDate,
                'currency'       => give_get_currency($form_id),
                'card_number'    => $cardNumber,
                'exp_month'      => $month,
                'exp_year'       => $year,
                'cvc'            => $cc_cvc,
                'cardholder_name' => $cc_holder,
                'payment_type'   => 'recurring',
                'interval'       => $data['period'],
                'allowed_cycles' => 12,
            ]),
            CURLOPT_HTTPHEADER => array(
                'X-Api-Key: doggiedaycareKxYeMhRl',
                'X-Api-Secret: doggiedaycareIBagbKnt',
                'Content-Type: application/json',
            ),
        ));
        $response = json_decode(curl_exec($curl), true);
        // Check for cURL errors
        if (curl_errno($curl)) {
            throw new Exception('cURL error: ' . curl_error($curl));
        }
        curl_close($curl);
        // Check if the charge ID is present in the response
        if (isset($response['charge_id'])) {
            $chargeId = $response['charge_id'];
            return $chargeId;
        } else {
            throw new Exception('Charge ID not found in the response.');
        }
         return array_merge([
             'success' => true,
            'transaction_id' => $response['charge_id'],
           'subscription_id' => $response['subscription_id'],
         ], $data);
    }
```
    
while I am going with One time Payment I am getting Error of "EasyMerchant Charge ID is required." And While going with Recrurring Error is "Sorry we're unable to process subscription donations with EasyMerchant. Please switch to the Stripe - Credit Card payment gateway to process a subscription donation."
    how to proceed with recurring?
    I am using easymerchant gateway api. But not sure, doing correct way or not. Can you guide me please.

**Reply by jonwaldstein** [reply]
@dightinfotechcom I should have mentioned, the initial PR I created did not include recurring support.  I was just trying to get you to a starting point for the gateway api and once your able to make a one-time donation successfully - then we can move on to recurring support.

Can you send me api docs for this gateway and if this gateway should be displaying credit card fields vs a redirect offsite?

**Reply by dightinfotechcom** [reply]
I tried what you suggested to me. But it didn't work for one time also. Where can I send the api docs? Can I send to your mail Id?? As you see the repo, there was a js link mentioned. You also add your js just below the line. In that js file, a pop up model is created and a credit card field is defined with authorisation. There is also a field for ACH offsite. I have sent you mail which is given at your GitHub profile. I have attached the api docs. Please look into it. I am getting an error for both one time and recurring. I think I am doing something wrong.

**Reply by jim-easymerchant** [reply]
@jonwaldstein - thank you for the help! generally we have two approaches to how we handle subscriptions with integrations:

Initial payment for subscription is made via direct post using our charge endpoint (https://developers.lyfepay.io/lyfepay?id=create_card_charge) OR hosted checkout. Saved card tokens are returned in the response. Recurring logic is handled within the integrator's app using the saved card token, and any additional logic related to the subscription including subsequent payments is handled by the integrator's application.

Initial payment for subscription is made via direct post using our charge endpoint with payment_type='recurring', and passing in the interval and allowed_cycles (https://developers.lyfepay.io/lyfepay?id=create_card_charge) ... here the subscription ID is returned, and we manage the recurring logic and schedule payments on our side. Additional subscription related endpoints for renewal, cancellation, etc... can be found here: https://developers.lyfepay.io/lyfepay?id=subscription_renew ... these endpoints are usually integrated into additional actions in the integrator's application.

@dightinfotechcom - could you please provide the response that you are getting back in your CURL request within the plugin?

**Reply by jonwaldstein** [reply]
@jim-easymerchant out of curiosity, what's the difference between [lyfepay](https://developers.lyfepay.io/) and [easymerchant](https://developers.easymerchant.io/)? I also saw there are links to "givewp" plugins but they are broken.  Any insight would be helpful.
- https://developers.lyfepay.io/lyfepay/user-guide?id=give-wp
- https://developers.easymerchant.io/easymerchant/user-guide?id=give-wp
